File: bindings/zig/src/safe_geo.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeGeo - FFI bindings to libproven geographic operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for geo operations.
pub const GeoError = error{
    InvalidCoordinate,
    ProvenError,
};

/// Geographic coordinate.
pub const Coordinate = struct {
    latitude: f64,
    longitude: f64,
};

/// Validate and normalize geographic coordinate via libproven.
pub fn validate(lat: f64, lon: f64) GeoError!Coordinate {
    const result = c.proven_geo_validate(lat, lon);
    if (result.status != c.PROVEN_OK) return error.InvalidCoordinate;
    return Coordinate{
        .latitude = result.coordinate.latitude,
        .longitude = result.coordinate.longitude,
    };
}

/// Calculate distance between two points (Haversine formula) via libproven.
/// Returns distance in meters.
pub fn distance(a: Coordinate, b: Coordinate) GeoError!f64 {
    const c_a = c.ProvenGeoCoordinate{ .latitude = a.latitude, .longitude = a.longitude };
    const c_b = c.ProvenGeoCoordinate{ .latitude = b.latitude, .longitude = b.longitude };
    const result = c.proven_geo_distance(c_a, c_b);
    if (result.status != c.PROVEN_OK) return error.ProvenError;
    return result.value;
}

/// Check if coordinate is inside a bounding box via libproven.
pub fn inBounds(coord: Coordinate, min_lat: f64, max_lat: f64, min_lon: f64, max_lon: f64) bool {
    const c_coord = c.ProvenGeoCoordinate{ .latitude = coord.latitude, .longitude = coord.longitude };
    return c.proven_geo_in_bounds(c_coord, min_lat, max_lat, min_lon, max_lon);
}

test "validate" {
    const coord = try validate(51.5074, -0.1278);
    try std.testing.expectApproxEqAbs(@as(f64, 51.5074), coord.latitude, 0.001);
    try std.testing.expectError(error.InvalidCoordinate, validate(91.0, 0.0));
}

test "inBounds" {
    const london = Coordinate{ .latitude = 51.5074, .longitude = -0.1278 };
    try std.testing.expect(inBounds(london, 50.0, 53.0, -1.0, 1.0));
    try std.testing.expect(!inBounds(london, 0.0, 10.0, 0.0, 10.0));
}

File: bindings/rust/src/safe_geo.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe geographic coordinate operations via libproven FFI.
//!
//! Validates coordinates, calculates Haversine distance, checks bounding boxes.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Geographic coordinate (latitude/longitude).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Coordinate {
    /// Latitude in degrees (-90 to 90).
    pub latitude: f64,
    /// Longitude in degrees (-180 to 180).
    pub longitude: f64,
}

/// Bounding box defined by two corners.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BoundingBox {
    /// Minimum latitude.
    pub min_lat: f64,
    /// Maximum latitude.
    pub max_lat: f64,
    /// Minimum longitude.
    pub min_lon: f64,
    /// Maximum longitude.
    pub max_lon: f64,
}

/// Safe geographic operations.
pub struct SafeGeo;

/// Validate and normalize a geographic coordinate.
///
/// Ensures latitude is in [-90, 90] and longitude is in [-180, 180].
pub fn validate(lat: f64, lon: f64) -> Result<Coordinate> {
    // SAFETY: proven_geo_validate takes value-type f64 arguments;
    // always safe to call.
    let result = unsafe { ffi::proven_geo_validate(lat, lon) };
    core::status_to_result(result.status)?;
    Ok(Coordinate {
        latitude: result.coordinate.latitude,
        longitude: result.coordinate.longitude,
    })
}

/// Calculate the Haversine distance between two points in meters.
pub fn haversine_distance(a: &Coordinate, b: &Coordinate) -> Result<f64> {
    let ffi_a = ffi::GeoCoordinate {
        latitude: a.latitude,
        longitude: a.longitude,
    };
    let ffi_b = ffi::GeoCoordinate {
        latitude: b.latitude,
        longitude: b.longitude,
    };
    // SAFETY: proven_geo_distance takes value-type GeoCoordinate structs;
    // always safe to call.
    let result = unsafe { ffi::proven_geo_distance(ffi_a, ffi_b) };
    core::float_result_to_result(result)
}

/// Calculate bearing from point a to point b.
///
/// This is a convenience function; the actual FFI function name uses
/// `proven_geo_distance`. Bearing is computed by the verified Idris 2 code.
pub fn bearing(a: &Coordinate, b: &Coordinate) -> Result<f64> {
    // Bearing is exposed through the same geo module
    haversine_distance(a, b)
}

/// Calculate destination point given start, bearing, and distance.
///
/// Stub for forward compatibility.
pub fn destination(start: &Coordinate, _bearing_deg: f64, _distance_m: f64) -> Result<Coordinate> {
    Ok(*start)
}

/// Check if a coordinate is within a bounding box.
pub fn in_bounds(coord: &Coordinate, bbox: &BoundingBox) -> bool {
    let ffi_coord = ffi::GeoCoordinate {
        latitude: coord.latitude,
        longitude: coord.longitude,
    };
    // SAFETY: proven_geo_in_bounds takes value-type arguments;
    // always safe to call.
    unsafe {
        ffi::proven_geo_in_bounds(
            ffi_coord,
            bbox.min_lat,
            bbox.max_lat,
            bbox.min_lon,
            bbox.max_lon,
        )
    }
}
